hazelcast/include/hazelcast/client/sql/impl/read_optimized_lru_cache.h

```diff
@@ -0,0 +1,233 @@
+            do_cleanup();
+        }
+    }
+
+    /**
+     * @param key the key of the cache entry
+     * Removes the cached value for the given key
+     */
+    void remove(const K& key) { cache_.remove(key); }
+
+protected:
+    /**
+     * Helper class to for simulation atomic lock with RAII
+     */
+    class custom_atomic_lock
+    {
+    public:
+        explicit custom_atomic_lock(bool initial_value = false)
+          : lock_(initial_value)
+        {
+        }
+
+        custom_atomic_lock(const custom_atomic_lock&) = delete;
+        const custom_atomic_lock& operator=(const custom_atomic_lock&) = delete;
+
+        bool try_lock()
+        {
+            bool expected = false;
+            return lock_.compare_exchange_strong(expected, true);
+        }
+
+        void release() { lock_.store(false); }
+
+        ~custom_atomic_lock() { release(); }
+
+    private:
+        std::atomic<bool> lock_;
+    };
+
+    /**
+     * Helper class to hold the value with timestamp.
+     */
+    template<typename T>
+    class value_and_timestamp
+    {
+    public:
+        const T value_;
+        int64_t timestamp_;
+
+        value_and_timestamp(T value)
+          : value_(value)
+        {
+            touch();
+        }
+
+        void touch() { timestamp_ = util::current_time_nanos(); }
+    };
+
+    util::SynchronizedMap<K, value_and_timestamp<V>> cache_;
+
+private:
+    /**
+     * Cleans the cache
+     */
+    void do_cleanup()
+    {
+        // if no thread is cleaning up, we'll do it
+        if (!cleanup_lock_.try_lock()) {
+            return;
+        }
+
+        try {
+            if (capacity_ >= cache_.size()) {
+                // this can happen if the cache is concurrently modified
+                return;
+            }
+            auto entries_to_remove = cache_.size() - capacity_;
+
+            /*max heap*/
+            std::priority_queue<int64_t> oldest_timestamps;
+
+            // 1st pass
+            const auto values = cache_.values();
+            for (const auto& value_and_timestamp : values) {
+                oldest_timestamps.push(value_and_timestamp->timestamp_);
+                if (oldest_timestamps.size() > entries_to_remove) {
+                    oldest_timestamps.pop();
+                }
+            }
+
+            // find out the highest value in the queue - the value, below which
+            // entries will be removed
+            if (oldest_timestamps.empty()) {
+                // this can happen if the cache is concurrently modified
+                return;
+            }
+            int64_t remove_threshold = oldest_timestamps.top();
+            oldest_timestamps.pop();
+
+            // 2nd pass
+            cache_.remove_values_if(
+              [remove_threshold](const value_and_timestamp<V>& v) -> bool {
+                  return (v.timestamp_ <= remove_threshold);
+              });
+
+            cleanup_lock_.release();
```

Comment:
`finally` class in `util` can be used here to execute `release` method at the end of scope.

Reply:
Fixed